### ISSUE TITLE
libjpeg-turbo is not possible to use on Android 6.0+ running x86 because of non-PIC code

### ISSUE THREAD

**Issue by Olof-IL** [libjpeg-turbo/libjpeg-turbo#155]
From android 23, the OS will refuse to load any shared library that contains TEXTRELS.

libjpeg-turbo has non-PIC code and is built with textrel warnings turned off.
This is from Android.mk file:
"# Turn off position independent code (PIC) warning.  This is because YASM"
"# cannot generate position independent code.  YASM is the build tool for the"
"# x86 and x86-64 SIMD."

This unfortunately means libjpeg turbo is not possible to use for any project that targets Android devices with x86 running Android 23 or above.

**Comment by dcommander** [comment]
It works fine on my Android device (Google Nexus 5X running 6.0.)  How do I reproduce the problem?  Also, it is unclear how you are building libjpeg-turbo or where exactly you're getting the error (Android.mk is a downstream build system-- not something we support.)

**Comment by alh42** [comment]
The nonPIC code is only in the x86 build of the library. 

Nexus5X has an ARM processor, so it has no problem, you need to compile it for x86 and try run it on an android device with x86. There are lots of Asus android devices with x86, and also Chromebooks, on those you will get unsatisfied link error, and and error message in adb logcat about the library containing text relocations, which are no longer supported for security reasons. (on earlier android you only get a warning)

I'll check how we were building the library, I assumed we used the Android.mk file, but I wasn't the one building it. I'll try to rebuild according to instructions in BUILDING.md and see if the problem still is there.

**Comment by Olof-IL** [comment]
The nonPIC code is only in the x86 build of the library.

Nexus5X has an ARM processor, so it has no problem, you need to compile it for x86 and try run it on an android device with x86. There are lots of Asus android devices with x86, and also Chromebooks, on those you will get unsatisfied link error, and and error message in adb logcat about the library containing text relocations, which are no longer supported for security reasons. (on earlier android you only get a warning)

I'll check how we were building the library, I assumed we used the Android.mk file, but I wasn't the one building it. I'll try to rebuild according to instructions in BUILDING.md and see if the problem still is there.

**Comment by dcommander** [comment]
I don't have any x86 Android devices.  I'll see if I can reproduce it using the emulator, though.

**Comment by Olof-IL** [comment]
A small update; I've managed to build a static library for x86 Android now, that works on 23+.

I first tried to follow the BULDING.md instructions, but I failed to get autoconf to generate a working make file. I couldn't get it to accept the x86 compiler in the Android toolchain. (I can provide logs if interested) but that could be partly because I'm a noob at autoconf.

I also tried to get the version published by google to work, but they've invented yet another build system, so I couldn't get it to work without the AOSP toolchain. (and downloading and setting up AOSP takes forever...)

But after a bit of tinkering with the cmake support in Android Studio, and the asm support in NDK r15, and borrowing the config from google, I managed to give the assembler the -DPIC flag, which resulted in a working library without text relocations

**Comment by dcommander** [comment]
BUILDING.md doesn't have x86 Android instructions, so that's probably where you ran into problems.  I'll need to try it myself and figure out the magic incantation, but what I can say is that the build system will pass `-DPIC` to NASM if you pass `--with-pic` to the configure script.  That might be the secret to success, here.

**Comment by Olof-IL** [comment]
That was indeed one of the problems, so I changed the script snipped in BUILDING.md to point out the x86 compiler in my NDK, but it didn't accept it as working. This is what I get in the config.log:

Thread model: posix
gcc version 4.9.x 20150123 (prerelease) (GCC) 
configure:3672: $? = 0
configure:3661: /Users/olle/Library/Android/sdk/ndk-bundle/toolchains/x86-4.9/prebuilt/darwin-x86_64/bin/i686-linux-android-gcc -V >&5
i686-linux-android-gcc: error: unrecognized command line option '-V'
i686-linux-android-gcc: fatal error: no input files
compilation terminated.
configure:3672: $? = 1
configure:3661: /Users/olle/Library/Android/sdk/ndk-bundle/toolchains/x86-4.9/prebuilt/darwin-x86_64/bin/i686-linux-android-gcc -qversion >&5
i686-linux-android-gcc: error: unrecognized command line option '-qversion'
i686-linux-android-gcc: fatal error: no input files
compilation terminated.
configure:3672: $? = 1
configure:3692: checking whether the C compiler works
configure:3714: /Users/olle/Library/Android/sdk/ndk-bundle/toolchains/x86-4.9/prebuilt/darwin-x86_64/bin/i686-linux-android-gcc -I/Users/olle/Library/Android/sdk/ndk-bundle/platforms/android-18/arch-arm64/usr/include -I/Users/olle/Library/Android/sdk/ndk-bundle/toolchains/x86-4.9/prebuilt/darwin-x86_64/include --sysroot=/Users/olle/Library/Android/sdk/ndk-bundle/platforms/android-18/arch-arm64 -O3 -fPIE -I/Users/olle/Library/Android/sdk/ndk-bundle/platforms/android-18/arch-arm64/usr/include -I/Users/olle/Library/Android/sdk/ndk-bundle/toolchains/x86-4.9/prebuilt/darwin-x86_64/include --sysroot=/Users/olle/Library/Android/sdk/ndk-bundle/platforms/android-18/arch-arm64 --sysroot=/Users/olle/Library/Android/sdk/ndk-bundle/platforms/android-18/arch-arm64 -pie conftest.c  >&5
/Users/olle/tools/android-sdk-macosx/ndk-bundle/toolchains/x86-4.9/prebuilt/darwin-x86_64/bin/../lib/gcc/i686-linux-android/4.9.x/../../../../i686-linux-android/bin/ld: error: cannot open crtbegin_dynamic.o: No such file or directory
/Users/olle/tools/android-sdk-macosx/ndk-bundle/toolchains/x86-4.9/prebuilt/darwin-x86_64/bin/../lib/gcc/i686-linux-android/4.9.x/../../../../i686-linux-android/bin/ld: error: cannot open crtend_android.o: No such file or directory
/Users/olle/tools/android-sdk-macosx/ndk-bundle/toolchains/x86-4.9/prebuilt/darwin-x86_64/bin/../lib/gcc/i686-linux-android/4.9.x/../../../../i686-linux-android/bin/ld: error: cannot find -lc
/Users/olle/tools/android-sdk-macosx/ndk-bundle/toolchains/x86-4.9/prebuilt/darwin-x86_64/bin/../lib/gcc/i686-linux-android/4.9.x/../../../../i686-linux-android/bin/ld: error: cannot find -ldl
collect2: error: ld returned 1 exit status
configure:3718: $? = 1
configure:3756: result: no
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "libjpeg-turbo"
| #define PACKAGE_TARNAME "libjpeg-turbo"
| #define PACKAGE_VERSION "1.5.2"
| #define PACKAGE_STRING "libjpeg-turbo 1.5.2"
| #define PACKAGE_BUGREPORT ""
| #define PACKAGE_URL ""
| #define PACKAGE "libjpeg-turbo"
| #define VERSION "1.5.2"
| /* end confdefs.h.  */
| 
| int
| main ()
| {
| 
|   ;
|   return 0;
| }
configure:3761: error: in `/Users/olle/Documents/git/libjpeg-turbo/build':
configure:3763: error: C compiler cannot create executables

**Comment by dcommander** [comment]
After experimenting with Android/x86 builds using a Nexus 5X/Android 23 AVD in the emulator, I now have a full understanding of the issue.  Here's "the deal":

* Android >= 16 only supports position-independent executables (those built with `-fPIE` and linked with `-pie`), which is why the Android build instructions in BUILDING.md specify those flags.  The executables simply won't run otherwise.
* Those flags don't automatically cause position-independent code to be used when assembling the x86 SIMD routines.  Our build system will only pass `-DPIC` to NASM if  `--with-pic` is passed to configure or if shared libraries are being built.  That probably explains why your initial build with Android.mk did not work.  That build system was not automatically passing `-DPIC` to NASM as our build system does.
* When `-DPIC` is not passed to NASM, using the libjpeg API through a static library produces a warning, and attempting to use the libjpeg API through a shared library produces the error you encountered.

As far as the compiler, I added Android/x86 build instructions to BUILDING.md and verified that they work.  The failure you were having above was due to the fact that you were setting the sys. root to **/Users/olle/Library/Android/sdk/ndk-bundle/platforms/android-18/arch-arm64** rather than **/Users/olle/Library/Android/sdk/ndk-bundle/platforms/android-18/arch-x86**.  Let me know if you are able to get a successful build using the new build instructions.
